mkfs: add read-after-write verification for main boot sector

anmuxixixi opened the pull request:
Recently, we encountered a confusing behavior during testing and would like to ask for your insight.

We have a batch of SD cards that fail to format on Windows, where the system reports:
“Windows was unable to complete the format.”

<img width="344" height="116" alt="image" src="https://github.com/user-attachments/assets/62e8e14c-fc67-4034-9333-5122f2567c03" />

**However, on the same devices, running mkfs.exfat on Linux completes successfully and reports no error.**

---

mkfs.exfat assumes that successful write system calls imply that data is eventually committed to persistent storage.

On some removable flash devices (e.g. low-quality SD cards, USB flash drives), this assumption does not hold. Such devices may:

- Acknowledge write requests without programming NAND ("fake write")
- Serve subsequent reads from internal controller cache

As a result, mkfs.exfat may report successful filesystem creation even though critical on-disk structures were never persisted.

Therefore, we would like to provide a warning for such SD cards, indicating that `mkfs.exfat` could not be completed successfully.

This patch adds a sanity check after writing the main boot sector to ensure the data can be read back correctly, improving robustness on devices with unreliable write behavior.

I have already tested this patch on the same SD card, and it now correctly reports an error indicating that formatting could not be completed.

anmuxixixi committed:
mkfs: add read-after-write verification for volume boot record

mkfs.exfat assumes that successful write system calls imply that data is
eventually committed to persistent storage.

On some removable flash devices (e.g. low-quality SD cards, USB flash
drives), this assumption does not hold. Such devices may:

- Acknowledge write requests without programming NAND ("fake write")
- Serve subsequent reads from internal controller cache

As a result, mkfs.exfat may report successful filesystem creation
even though critical on-disk structures were never persisted.

This patch adds a sanity check after writing the volume boot record to
ensure the data can be read back correctly, improving robustness on
devices with unreliable write behavior.

Signed-off-by: YangWen <anmuxixixi@gmail.com>

namjaejeon commented:
@anmuxixixi Can you run mkfs.ext4 or mkfs.xfs on your broken storage and tell me what the result is ?

anmuxixixi commented:
> @anmuxixixi Can you run mkfs.ext4 or mkfs.xfs on your broken storage and tell me what the result is ?

I have just formatted the drive using mkfs.ext4. The command reported a successful execution.

```shell
amx@anmuxixixi $ mkfs.ext4 /dev/block/sda
mke2fs 1.45.4
/dev/block/sda contains a vfat file system
Proceed anyway? (y,N) y
# ...

Allocating group tables: done
Writing inode tables: done
Creating journal (xxxx blocks)
done
Writing superblocks and filesystem accounting information: done
```

Currently, our mobile devices use mkfs.exfat. It reports success even when the subsequent mount fails, which confuses users.I’m seeking your professional advice。

I also noticed the validation logic in badblocks within the e2fsprogs repo. So, We’re thinking of implementing something similar to verify disk integrity post-formatting.

anmuxixixi commented:
@namjaejeon Hi. I believe this change enhances the user experience without impacting performance.
If you think there’s a better way to achieve this, please let me know and I’ll be happy to refine the implementation.
Thanks.

namjaejeon commented:
@anmuxixixi 

I have three questions.
1. Can this method find perfectly the broken device since it doesn't validate any metadata other than the boot sector?
2. Other Linux format tools don't have validation code like this. Including validation could degrade performance, but is there any reason to degrade performance under normal circumstances for this extremely small possibility?
3. Can't you just validate with fsck.exfat after mkfs?

So, If you add new option to include the validation in mkfs, I can accept it because it doesn't affect other users. And if you are going to add the new option, you have to answer my first question.

namjaejeon commented:
@anmuxixixi 
Are you interested in this issue?

https://github.com/exfatprogs/exfatprogs/issues/261

mkfs.fat support this, so we might want to refer to dosfstools... (https://github.com/dosfstools/dosfstools/blob/master/src/mkfs.fat.c#L2010)

anmuxixixi commented:
> @anmuxixixi Are you interested in this issue?
> 
> #261
> 
> mkfs.fat support this, so we might want to refer to dosfstools... (https://github.com/dosfstools/dosfstools/blob/master/src/mkfs.fat.c#L2010)

Thanks for your guidance.I'll check it.

namjaejeon commented:
@anmuxixixi dosfstools is licensed by GPLv3. So we don't copy codes from this. just refer how to handle it...
And Please add the new option to include validation in mkfs.
Thanks for your works!

anmuxixixi commented:
Recently, while reading the code of dosfstools, I noticed that its approach is quite similar to the badblocks tool in the e2fsprogs repository. During formatting, dosfstools only performs a read-only test over all blocks and does not perform any write test.

I also ran a small experiment on my 128 GB SD card using badblocks:

* `badblocks -s -v /dev/block/sda`  cost **76 minutes**

* `badblocks -s -v -w /dev/block/sda`  cost **358 minutes (about 6 hours)**

This made me think that if a similar read-only test were added to mkfs.exfat, encountering a large-capacity SD card (for example, 2 TB) could easily result in a very frustrating user experience.

In addition, I captured the SCSI traffic during formatting on Windows using Wireshark, and observed the following sequence of operations.


| SCSI Cmd | Start Sector | Sectors | Region      |
|----------|--------------|---------|-------------|
| WRITE(10)| 0            | 24      | Boot record |
| READ(10) | 0            | 24      | Boot record |
| WRITE(10)| 64           | 1       | N/A         |
| READ(10) | 64           | 1       | N/A         |
| WRITE(10)| 2048         | 1       | FAT         |
| READ(10) | 2048         | 1       | FAT         |
| WRITE(10)| 6144         | 1       | Bitmap      |
| READ(10) | 6144         | 1       | Bitmap      |
| WRITE(10) | 6145| 128      | Bitmap      |
| WRITE(10) | 6273| 128      | Bitmap      |
|  READ(10)  | 6145| 128      | Bitmap      |
|  READ(10) | 6273| 128      | Bitmap      |
| WRITE(10) | 6400| 12| upcase|
|  READ(10)  | 6400| 12| upcase|
| WRITE(10) | 6566| 128| root dir|
|  WRITE(10) | 6784| 128| upcase|
| READ(10)| 6566| 128| root dir|
|  READ(10) | 6784| 128| root dir||

> Although I am not sure why Windows writes to sector 64th during formatting, this behavior was consistently observed in my SCSI trace


Based on this behavior, I modified my patch that more closely follows the Windows formatting sequence, and I also added an additional option to control this behavior during formatting.

namjaejeon commented on mkfs/mkfs.c:
```diff
@@ -39,6 +39,48 @@ static unsigned int get_new_serial(void)
+
+	if (memcmp(buf, verify, len) != 0) {
+		exfat_debug("%s verify mismatch (off=%llu)\n",
+			  what, (unsigned long long)off);
```
please add ret = -EIO;

namjaejeon commented on mkfs/mkfs.c:
```diff
@@ -39,6 +39,48 @@ static unsigned int get_new_serial(void)
+	n = exfat_read(fd, verify, len, off);
+	if (n != (ssize_t)len) {
+		exfat_debug("%s verify read failed (off=%llu)\n",
+			  what, (unsigned long long)off);
```
please add ret = -EIO;

namjaejeon commented on mkfs/mkfs.c:
```diff
@@ -39,6 +39,48 @@ static unsigned int get_new_serial(void)
+	}
+
+	free(verify);
+	return 0;
```
please use ret instead of return 0;

namjaejeon commented on mkfs/mkfs.c:
```diff
@@ -39,6 +39,48 @@ static unsigned int get_new_serial(void)
+	return 0;
+
+err:
+	free(verify);
```
please remove goto err routine.

namjaejeon commented on mkfs/mkfs.c:
```diff
@@ -573,20 +644,64 @@ static int make_exfat(struct exfat_blk_dev *bd, struct exfat_user_input *ui)
+			free(boot_secs);
+			return -EIO;
+		}
+
```
You read boot sector again here. cache data of boot sector could be invalidate by reclaim or drop cache. So we can read corrupted data from broken device. We need to compare checksum of boot sector after reading it.

anmuxixixi replied:
Thank you for your detailed feedback and guidance.

In my implementation, I read back all 24 sectors covering both the main and backup boot sectors, and then perform a memcmp against the data I just wrote. Since the boot sector checksum is contained within these sectors, this effectively verifies the checksum as well.

My intention here is to confirm that the data has truly been written to the device. If there is any aspect I may have misunderstood or if there are edge cases I have overlooked, I would greatly appreciate your guidance on how to improve this verification.

namjaejeon replied:
I know, but my point is that you're re-reading the boot sector with new buffer, not using the write buffer, and comparing it against it. In a memory-poor environment, this validation becomes meaningless if you're reading from a broken disk, not from cached data in memory via a page cache hit.

namjaejeon commented on mkfs/mkfs.c:
```diff
@@ -573,20 +644,64 @@ static int make_exfat(struct exfat_blk_dev *bd, struct exfat_user_input *ui)
+			free(fat_sec);
+			return -EIO;
+		}
+
```
This seems like a difficult part to properly validate. Should we just validate the sector size of the fat table? Wouldn't the fat table be larger than 512 bytes on devices with 2TB or more?

namjaejeon replied:
When reading from here, how do you guarantee that this buffer was read from the page cache? If your system runs low on memory, the page caches can be dropped.

anmuxixixi replied:
> When reading from here, how do you guarantee that this buffer was read from the page cache? If your system runs low on memory, the page caches can be dropped.

The intent here is to ensure the verification read comes from the physical device.
After writing, I call fsync() to flush dirty pages, then explicitly invalidate the block device page cache using BLKFLSBUF before re-reading, so the subsequent read is forced to hit the device and not page cache.

If there is any misunderstanding or if this approach is incorrect, please let me know. Thanks.

namjaejeon replied:
I am not saying exfat_check_written_data(), I am pointing out that you re-read fat table with fat_sec buffer and pass it to exfat_check_written_data().  You don't expect if fat_sec is data you are written.

anmuxixixi replied:
Thanks for your reminder and guidance. I’ve updated the patch accordingly.

namjaejeon commented on mkfs/mkfs.c:
```diff
@@ -270,48 +322,75 @@ static int exfat_create_fat_table(struct exfat_blk_dev *bd,
 	clu = write_fat_entries(ui, bd->dev_fd, EXFAT_FIRST_CLUSTER,
 		finfo.bitmap_byte_len);
 	if (clu < 0)
-		return ret;
```
return 0 can be returned although write_fat_entries fails.
ret = clu;

namjaejeon commented on defrag/defrag.c:
```diff
@@ -136,7 +136,8 @@ static int exfat_rw_bitmap(struct exfat *exfat, bool read)
  * Updates the boot sector checksum.
  */
-static int exfat_update_boot_checksum(struct exfat_blk_dev *bd, bool is_backup)
+static int exfat_update_boot_checksum(struct exfat_blk_dev *bd,
```
I don't like to add ui argument here. you can check if ui is null or not in exfat_write_checksum_sector().

namjaejeon replied:
you need to remove ui arguments in other functions.

namjaejeon commented on defrag/defrag.c:
```diff
@@ -167,7 +168,7 @@ static int exfat_update_boot_checksum(struct exfat_blk_dev *bd, bool is_backup)
 	}
 
-	ret = exfat_write_checksum_sector(bd, checksum, is_backup);
+	ret = exfat_write_checksum_sector(bd, ui, checksum, is_backup);
```
So you can give null for ui argument here.
ret = exfat_write_checksum_sector(bd, NULL, checksum, is_backup);

anmuxixixi replied:
This is a perfect modification, thank you for your guidance.

namjaejeon commented on lib/libexfat.c:
```diff
@@ -738,6 +777,17 @@ int exfat_write_checksum_sector(struct exfat_blk_dev *bd,
 		goto free;
 	}
 
+	if (ui->verify) {
```
if (ui && ui->verify) {

namjaejeon commented on lib/libexfat.c:
```diff
@@ -684,6 +684,44 @@ int exfat_set_volume_guid(struct exfat *exfat, const char *guid)
+		return ret;
+
+#ifdef BLKFLSBUF
+	if (ioctl(fd, BLKFLSBUF, 0) != 0)
```
I am concerned about the potential side effects of BLKFLSBUF. Since it has a global impact on the block device, could calling this ioctl on one partition (e.g., /dev/sda1) inadvertently invalidate the cache or cause I/O stalls for other active partitions (e.g., /dev/sda2) on the same physical hardware?

namjaejeon replied:
Using Direct I/O (O_DIRECT) could be a better alternative.

anmuxixixi replied:
Okay, I have added a new verify_fd and implemented the use of O_DIRECT to open it.

namjaejeon commented on mkfs/mkfs.c:
```diff
@@ -428,6 +531,7 @@ static const struct option opts[] = {
 	{"boundary-align",	required_argument,	NULL,	'b' },
 	{"pack-bitmap",		no_argument,		NULL,	PACK_BITMAP },
 	{"full-format",		no_argument,		NULL,	'f' },
+	{"verify-written",	no_argument,		NULL,	'C' },
```
Why -C option ? it is not matched with verify-written.
how about "check-written" with -c option ?

anmuxixixi replied:
The option name has been changed from -verify-write to -check-written.
-c option is already taken for the cluster size.

namjaejeon commented on lib/libexfat.c:
```diff
@@ -684,6 +697,47 @@ int exfat_set_volume_guid(struct exfat *exfat, const char *guid)
+	if (ret)
+		return ret;
+
+	posix_fadvise(bd->dev_fd, off, len, POSIX_FADV_DONTNEED);
```
Is it needed although you use direct-I/O ?

anmuxixixi replied:
I have deleted it. And i conducted a test on the same SD card, when i run mkfs.exfat, it reported failed.
Sorry.I realized that I accidentally removed the fsync.I'll re-send a new commit tonight.Thanks~

namjaejeon commented on mkfs/mkfs.c:
```diff
@@ -24,6 +24,7 @@
 #include "mkfs.h"
 
 struct exfat_mkfs_info finfo;
+unsigned char *fat_table_buff;
```
can we add fat_table_buff into ui ? (i.e. ui->fat_table_buff)

namjaejeon replied:
@anmuxixixi I didn't get the answer about the above from you yet.

namjaejeon replied:
@anmuxixixi Ah, I have checked that you updated it now. Thanks.

namjaejeon commented on mkfs/mkfs.c:
```diff
@@ -428,6 +537,7 @@ static const struct option opts[] = {
 	{"boundary-align",	required_argument,	NULL,	'b' },
 	{"pack-bitmap",		no_argument,		NULL,	PACK_BITMAP },
 	{"full-format",		no_argument,		NULL,	'f' },
+	{"check-written",	no_argument,		NULL,	'C' },
```
Is there any reason why you use 'C' instead of 'c' ?

anmuxixixi replied:
-c option is already taken for the cluster size.

namjaejeon commented:
Applied it to #exfat-next now.
Thanks!
